src/Take.Blip.Builder/Actions/ExecuteScriptV2/ExecuteScriptV2Action.cs

```diff
@@ -0,0 +1,178 @@
+﻿using System;
+using System.ComponentModel.DataAnnotations;
+using System.Diagnostics;
+using System.Threading;
+using System.Threading.Tasks;
+using Lime.Protocol;
+using Microsoft.ClearScript;
+using Microsoft.ClearScript.V8;
+using Serilog;
+using Take.Blip.Builder.Actions.ExecuteScriptV2.Functions;
+using Take.Blip.Builder.Hosting;
+using Take.Blip.Builder.Utils;
+
+namespace Take.Blip.Builder.Actions.ExecuteScriptV2
+{
+    /// <inheritdoc />
+    public class ExecuteScriptV2Action : ActionBase<ExecuteScriptV2Settings>
+    {
+        private readonly IConfiguration _configuration;
+        private readonly IHttpClient _httpClient;
+        private readonly ILogger _logger;
+
+        /// <inheritdoc />
+        public ExecuteScriptV2Action(IConfiguration configuration, IHttpClient httpClient,
+            ILogger logger)
+            : base(nameof(ExecuteScriptV2))
+        {
+            HostSettings.CustomAttributeLoader = new LowerCaseMembersLoader();
+
+            _configuration = configuration;
+            _httpClient = httpClient;
+            _logger = logger;
+        }
+
+        /// <inheritdoc />
+        public override async Task ExecuteAsync(IContext context, ExecuteScriptV2Settings settings,
+            CancellationToken cancellationToken)
+        {
+            try
+            {
+                var arguments = await GetScriptArgumentsAsync(context, settings, cancellationToken);
+
+                using var engine = new V8ScriptEngine(
```

Comment:
No, we can't...

The script engine is not thread safe. Think the engine as a browser window. If we kept a single instance of the engine, each user would be able to access the environment and also the functions (or global variables) defined by other users. Everything would colide, for example functions with same names, etc.

We could make this better by keeping a single engine for each's bot action and sharing only with the same bot and the same action, but I don't think this is possible with the current state of how things are implemented. We could, maybe, keep a cache of engines but I think this can get complex very fast.

I really think the cache would improve the execution a lot by not creating the engine and declaring everything each execution, but we would also create race-conditions to our users. For example if I create a global variable inside my script, the global variable value from previous executions would be kept to next executions.

To do it we would need to make possible for our users to define a "tearDown" function, for example, for them to clear previous global states.

But this also increases the overhead to the engine itself. With the current version, parallel executions are possible because each execution do not share state with other executions (a new engine each time a script is called). If we share the engine, V8 limits (for example for parallel threads) would apply and maybe we could make everything worse.

I think there may be opportunities for improvements, but they should be implemented with care and based entirelly on benchmarks. It is something we can look into in the future.